010-provider-architecture/002-final/lib/core/viewmodels/comments_model.dart

```diff
@@ -1,18 +1,35 @@
 import 'package:provider_architecutre/core/enums/viewstate.dart';
 import 'package:provider_architecutre/core/models/comment.dart';
 import 'package:provider_architecutre/core/services/api.dart';
-
-import '../../locator.dart';
 import 'base_model.dart';
 
 class CommentsModel extends BaseModel {
-  Api _api = locator<Api>();
+  Api _api;
+  Api get api => _api;
+  set api(Api api) {
+    _api = api;
+    notifyListeners();
+  }
+
+  int _postId;
+  int get postId => _postId;
+  set postId(int postId) {
+    if (postId != _postId) {
+      _postId = postId;
+      _fetchComments(postId);
+      notifyListeners();
+    }
+  }
 
-  List<Comment> comments;
+  List<Comment> _comments;
+  List<Comment> get comments => _comments;
 
-  Future fetchComments(int postId) async {
+  Future _fetchComments(int postId) async {
+    if (state == ViewState.Busy) {
```

Comment:
I don't agree with any of the Changes in the Comments Model. Especially the fact that we would now have to check the busy state and throw errors because the code might already be running. Making the fetch "automatic" in some cases would be better. But not here, I am a proponent of readable code. I know that's not a big topic, but because I write code bases for other companies it needs to be easy to understand and most importantly easy to teach. New developers coming in shouldn't have a hard time at all just looking at the code and seeing what's happening. Which is why I had so little code in this class, I don't think it should be responsible for fetching it's own data. It needs to be told what to do, from my point of view and what I want from the architecture,

Reply:
Weren't you saying that you don't want business logic in your widgets?
Knowing when to catch is business logic. 

Similarly your previous code had a bug, where if you modified the postId from the comment page itself — it wouldn't load the new comments.
With this code, that bug never happens. 

The throw is unnecessary, but I kept your busy thing so I felt it was natural.

Reply:
Yes I did. When I say widget I mean the file that you describe as meta data to the UI and to which I refer to as the UI code. All my business logic is in my Model for the widget, not the widget itself. The bug wasn't there because I was using get_it over provider, it was just a bug based on the tutorial code. My tutorial is always the minimum amount of code to get my idea across and to teach the concept at a higher level.

Reply:
I'm not saying that the bug was there because of `get_it` :)
Just that this architecture fixed it without even trying to.